content/docs/rfcs/64/README.md

```diff
@@ -0,0 +1,258 @@
+
+### Message Attributes
+
+- The mandatory `payload` attribute MUST contain the message data payload as crafted by the application.
+- The mandatory `content_topic` attribute MUST specify a string identifier that can be used for content-based filtering. This is also crafted by the application. See dhuhfh for more on the content topic format.    
+- The optional `meta` attribute MAY be omitted. If present this will form part of the message uniqueness vector described in [14/WAKU2-MESSAGE](https://rfc.vac.dev/spec/14/).
+- The optional `version` attribute SHOULD be set to `0`. It MUST be interpreted as `0` if not present.
+- The mandatory `timestamp` attribute MUST contain the Unix epoch time at which the message was generated by the application. The value MUST be in nanoseconds. It MAY contain a fudge factor of up to 1 seconds in either direction to improve resistance to timing attacks.
+- The optional `ephemeral` attribute MUST be set to `true` if the message should not be persisted by the Waku Network.
+- The optional `rate_limit_proof` attribute SHOULD be populated with the RLN proof as set out in dhjfhjhd. Messages with this field unpopulated MAY be discarded from the network by relayers.
+
+### Message Size
+
+Any Waku Message published to the network MUST NOT exceed an absolute maximum size of `150` kilobytes, attributes included.
+The average size of Waku Messages publised by each publisher SHOULD NOT exceed `3` kilobytes.
+
+### Message Validation
+
+Relay nodes MUST apply [gossipsub v1.1 validation](https://github.com/libp2p/specs/blob/c96c9ec5909d64fe020d7630f3fd982bc18fd06a/pubsub/gossipsub/gossipsub-v1.1.md#extended-validators) to each relayed message and
+SHOULD apply all of the rules set out in the section below to determine the validity of a message.
+Validation has one of three outcomes,
+repeated here from the [gossipsub specification](https://github.com/libp2p/specs/blob/c96c9ec5909d64fe020d7630f3fd982bc18fd06a/pubsub/gossipsub/gossipsub-v1.1.md#extended-validators) for ease of reference:
+1. Accept - the message is considered valid and it MUST be delivered and forwarded to the network
+2. Reject - the message is considered invalid, MUST be rejected and SHOULD trigger a gossipsub scoring penalty against the transmitting peer
+3. Ignore - the message SHOULD NOT be delivered and forwarded to the network, but this MUST NOT trigger a gossipsub scoring penalty against the transmitting peer.
+
+The following validation rules are defined:
+
+#### Decoding failure
+
+If a message fails to decode as a valid [14/WAKU2-MESSAGE](https://rfc.vac.dev/spec/14/).
+the relay node MUST _reject_ the message.
+This SHOULD trigger a penalty against the transmitting peer.
+
+#### Invalid timestamp
+
+If a message has a timestamp deviating by more than `20` seconds
+either into the past or the future
+when compared to the relay node's internal clock,
+the relay node MUST _ignore_ the message.
```

Comment:
This may lead to DOS'ing, should we define a rate-limit for ignored messages which should eventually lead to peer disconnection if it continues?

Reply:
I don't see this as a DOS threat as the messages are discarded directly and not relayed, unless I'm missing something :)

Reply:
DOS attack is on the node receiving and discarding the messages. 
e.g a node can just keep replaying same old messages at a very high rate which can cause the node's resources to get choked reading, decoding and then ignoring these messages. 

Hence, we should have some sort of a limit or rate limit on ignored messages, based on which peer can be disconnected.

Reply:
Indeed. There are a number of direct DoS attacks on nodes that fall beyond the scope of the raw version of this spec to address. Nodes can also be DoSed by service requests or even API calls (if accessible). These should be addressed, but our first step is to define a minimal vector to protect the network. We'll augment this with more and more sophisticated mitigation models for specific attack vectors.

Reply:
Maybe a good start to take this approach for some ignore scenarios. 

But we will have to deal with ignored messages in the near-term as it would become an attack vector. 
At the same time, converting all ignore conditions into reject conditions may not be a good idea as the reasoning behind making them ignore would be lost (like slow peers or one-off cases).

Reply:
Ok, I see your point. My thinking was that we may not want to be more forgiving towards nodes for drifting clocks or experiencing processing delays, but I think it's safer to be more strict here and prefer nodes that can keep up. Changed in [637b5ac](https://github.com/vacp2p/rfc/pull/616/commits/637b5ac30d76f2e4c0f35f14cfa870d4682f2df5)